sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt

```diff
@@ -27,24 +29,85 @@ class QueueIntegrationTests : BaseTest() {
     }
 
     @Test
-    fun givenRunQueueAndFailTasksThenRerunQueue_expectQueueRerunsAllTasksAgain(): Unit = runBlocking {
+    fun givenRunQueueAndFailTasksThenRerunQueue_expectQueueRerunsAllTasksAgain(): Unit =
+        runBlocking {
+            val givenIdentifier = String.random
+            queue.queueIdentifyProfile(givenIdentifier, null, emptyMap())
+            queue.queueTrack(givenIdentifier, String.random, EventType.event, emptyMap())
+            queueStorage.getInventory().count() shouldBeEqualTo 2
+
+            mockWebServer.enqueueNoInternetConnection()
+            queue.run()
+
+            queueStorage.getInventory().count() shouldBeEqualTo 2
+            mockWebServer.requestCount shouldBeEqualTo 1
+
+            mockWebServer.enqueue(MockResponse().setResponseCode(200))
+            mockWebServer.enqueue(MockResponse().setResponseCode(200))
+            queue.run()
+
+            // expect all of tasks to run and run successfully
+            queueStorage.getInventory().count() shouldBeEqualTo 0
+            mockWebServer.requestCount shouldBeEqualTo 3
+        }
+
+    @Test
+    fun givenRunQueueAndFailWith400_expectAllGroupTasksToBeDeleted(): Unit = runBlocking {
         val givenIdentifier = String.random
         queue.queueIdentifyProfile(givenIdentifier, null, emptyMap())
         queue.queueTrack(givenIdentifier, String.random, EventType.event, emptyMap())
         queueStorage.getInventory().count() shouldBeEqualTo 2
 
-        mockWebServer.enqueueNoInternetConnection()
+        mockWebServer.enqueue(MockResponse().setResponseCode(400))
         queue.run()
 
-        queueStorage.getInventory().count() shouldBeEqualTo 2
+        // expect tasks to be deleted
+        queueStorage.getInventory().count() shouldBeEqualTo 0
         mockWebServer.requestCount shouldBeEqualTo 1
+    }
 
+    @Test
+    fun givenRunQueueAndFailWith400_expectNonGroupTasksNotToBeDeleted(): Unit = runBlocking {
+        val givenIdentifier = String.random
+        queue.queueIdentifyProfile(givenIdentifier, null, emptyMap())
+        queueStorage.getInventory().count() shouldBeEqualTo 1
         mockWebServer.enqueue(MockResponse().setResponseCode(200))
-        mockWebServer.enqueue(MockResponse().setResponseCode(200))
+
+        queue.queueTrack(givenIdentifier, String.random, EventType.event, emptyMap())
+        mockWebServer.enqueue(MockResponse().setResponseCode(400))
+
+        queue.queueTrack(givenIdentifier, String.random, EventType.event, emptyMap())
+        mockWebServer.enqueue(MockResponse().setResponseCode(404))
+
         queue.run()
 
-        // expect all of tasks to run and run successfully
-        queueStorage.getInventory().count() shouldBeEqualTo 0
+        // expect tasks with 400 still present
+        queueStorage.getInventory().count() shouldBeEqualTo 1
```

Comment:
I think it would be good to check that the 1 item remaining in the queue is the item we expect to see (the 404 response). 

You have done a lot of work on this PR. I'll help with this task.